Add docstrings for save_on_cpu hooks

Varal7 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62410

This PR adds docstrings for the `save_on_cpu` context-manager introduced in #61928.

Also uncomments the warning about pinned memory in CUDA semantics docs.

Depends on: #62361.

Differential Revision: [D29990129](https://our.internmc.facebook.com/intern/diff/D29990129)

Varal7 committed:
Add docstrings for save_on_cpu hooks

This PR adds docstrings for CPU hooks introduced in #61928.

Also uncomments the warning about pinned memory in CUDA semantics docs

[ghstack-poisoned]

Varal7 committed:
Update on "Add docstrings for save_on_cpu hooks"


This PR adds docstrings for CPU hooks introduced in #61928.

Also uncomments the warning about pinned memory in CUDA semantics docs.

Depends on: #62361.

For now docstrings are an orphan page at https://docs-preview.pytorch.org/62410/generated/torch.autograd.graph.set_save_on_cpu_hooks.html#torch-autograd-graph-set-save-on-cpu-hooks

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29990129).

Varal7 committed:
Update on "Add docstrings for save_on_cpu hooks"


This PR adds docstrings for CPU hooks introduced in #61928.

Also uncomments the warning about pinned memory in CUDA semantics docs.

Depends on: #62361.

For now docstrings are an orphan page at https://docs-preview.pytorch.org/62410/generated/torch.autograd.graph.set_save_on_cpu_hooks.html#torch-autograd-graph-set-save-on-cpu-hooks

Differential Revision: [D29990129](https://our.internmc.facebook.com/intern/diff/D29990129)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29990129).

Varal7 committed:
Update on "Add docstrings for save_on_cpu hooks"


This PR adds docstrings for CPU hooks introduced in #61928.

Also uncomments the warning about pinned memory in CUDA semantics docs.

Depends on: #62361.

For now docstrings are an orphan page at https://docs-preview.pytorch.org/62410/generated/torch.autograd.graph.set_save_on_cpu_hooks.html#torch-autograd-graph-set-save-on-cpu-hooks

Differential Revision: [D29990129](https://our.internmc.facebook.com/intern/diff/D29990129)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29990129).

Varal7 committed:
Update on "Add docstrings for save_on_cpu hooks"


This PR adds docstrings for CPU hooks introduced in #61928.

Also uncomments the warning about pinned memory in CUDA semantics docs.

Depends on: #62361.

For now docstrings are an orphan page at https://docs-preview.pytorch.org/62410/generated/torch.autograd.graph.set_save_on_cpu_hooks.html#torch-autograd-graph-set-save-on-cpu-hooks

Differential Revision: [D29990129](https://our.internmc.facebook.com/intern/diff/D29990129)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29990129).

Varal7 committed:
Update on "Add docstrings for save_on_cpu hooks"


This PR adds docstrings for CPU hooks introduced in #61928.

Also uncomments the warning about pinned memory in CUDA semantics docs.

Depends on: #62361.

For now docstrings are an orphan page at https://docs-preview.pytorch.org/62410/generated/torch.autograd.graph.set_save_on_cpu_hooks.html#torch-autograd-graph-set-save-on-cpu-hooks

Differential Revision: [D29990129](https://our.internmc.facebook.com/intern/diff/D29990129)

[ghstack-poisoned]

zou3519 commented on docs/source/notes/cuda.rst:
```diff
@@ -468,7 +468,7 @@ also preserve :class:`torch.device` and :class:`torch.dtype` of a Tensor).
-.. warning:
+.. warning::
 
     This is an advanced tip. You overuse of pinned memory can cause serious
```
Could we fix the grammar of "You overuse of pinned memory can cause serious" while we're at it? I think this is supposed to be "Your overuse of pinned memory" but maybe we can just write "If you overuse pinned memory this can cause problems if you are low on RAM"

zou3519 commented on torch/autograd/saved_variable_default_hooks.py:
```diff
@@ -8,18 +8,49 @@ def reset_saved_tensors_default_hooks():
+    """"Context-manager under which tensors saved by the forward pass will be
+    stored on cpu, then retrieved for backward.
+
+    When performing operations within this context manager, intermediary
+    results saved in the graph during the forward pass will be moved to CPU,
```
What do you think about a quick pitch on why this feature is useful? Something like "this may be helpful if your model doesn't fit in GPU memory"

zou3519 commented on torch/autograd/saved_variable_default_hooks.py:
```diff
@@ -8,18 +8,49 @@ def reset_saved_tensors_default_hooks():
+
+    Example::
+
+        >>> a = torch.randn(5, requires_grad=True, device="cuda")
+        >>> b = torch.randn(5, requires_grad=True, device="cuda")
+        >>> c = torch.randn(5, requires_grad=True, device="cuda")
+        >>> d = a * b # a and b are saved in the graph (on GPU)
+        >>> with torch.autograd.graph.save_on_cpu():
+        ...    e = d * c # d and c are saved on CPU
+        >>> f = a * e # a and e are saved on GPU
+        >>> del a, b, c, d, e
+        >>> # the content of a, b, e are still alive on GPU
+        >>> # the content of c and d only live on CPU
```
Might be good to have a .backward() somewhere to demonstrate that the backward pass still works and that the autograd computation graph is being moved to cpu

Also I found this example a bit difficult to read but I'm not sure what a better example would be

Varal7 replied:
Hopefully, this is more readable:

```python
 a = torch.randn(5, requires_grad=True, device="cuda")
 b = torch.randn(5, requires_grad=True, device="cuda")
 c = torch.randn(5, requires_grad=True, device="cuda")

def f(a, b, c):
    prod_1 = a * b           # a and b are saved on GPU
    with torch.autograd.graph.save_on_cpu():
        prod_2 = prod_1 * c  # prod_1 and c are saved on CPU
    y = prod_2 * a           # prod_2 and a are saved on GPU
    return y
    
y = f(a, b, c)
del a, b, c  # for illustration only 
# the content of a, b, and prod_2 are still alive on GPU
# the content of prod_1 and c only live on CPU
y.sum().backward()  # all CPU tensors are moved back to GPU
```

zou3519 replied:
Yes, that looks better to me, thanks for taking the effort to make it easier to read!

Varal7 committed:
Update on "Add docstrings for save_on_cpu hooks"


This PR adds docstrings for CPU hooks introduced in #61928.

Also uncomments the warning about pinned memory in CUDA semantics docs.

Depends on: #62361.

For now docstrings are an orphan page at https://docs-preview.pytorch.org/62410/generated/torch.autograd.graph.set_save_on_cpu_hooks.html#torch-autograd-graph-set-save-on-cpu-hooks

Differential Revision: [D29990129](https://our.internmc.facebook.com/intern/diff/D29990129)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29990129).

Varal7 committed:
Update on "Add docstrings for save_on_cpu hooks"


This PR adds docstrings for CPU hooks introduced in #61928.

Also uncomments the warning about pinned memory in CUDA semantics docs.

Depends on: #62361.

For now docstrings are an orphan page at https://docs-preview.pytorch.org/62410/generated/torch.autograd.graph.set_save_on_cpu_hooks.html#torch-autograd-graph-set-save-on-cpu-hooks

Differential Revision: [D29990129](https://our.internmc.facebook.com/intern/diff/D29990129)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29990129).

Varal7 committed:
Update on "Add docstrings for save_on_cpu hooks"


This PR adds docstrings for CPU hooks introduced in #61928.

Also uncomments the warning about pinned memory in CUDA semantics docs.

Depends on: #62361.

For now docstrings are an orphan page at https://docs-preview.pytorch.org/62410/generated/torch.autograd.graph.set_save_on_cpu_hooks.html#torch-autograd-graph-set-save-on-cpu-hooks

Differential Revision: [D29990129](https://our.internmc.facebook.com/intern/diff/D29990129)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29990129).
